readme.md
## Scheduling algorithm for planning tasks with day-based time quant

[![npm version](https://img.shields.io/npm/v/project-tasks-scheduling-engine?color=green&label=npm%20version)](https://www.npmjs.com/package/project-tasks-scheduling-engine)

Publications:
* In English — https://medium.com/@sky2high/task-planning-algorithm-in-typescript-53a20cef6163.
* На русском — https://habr.com/ru/post/515554/

## Install

```shell
npm install project-tasks-scheduling-engine
```

## Use

```typescript
import { scheduleTasks } from "project-tasks-scheduling-engine";

const todayForTest = new Date("2020-01-01");

console.log(
  scheduleTasks(
    [
      {
        id: "0",
        title: "Make scheduling algorithm",
        start: todayForTest,
        end: new Date("2020-01-04"),
        duration: 4,
        position: 0,
        progress: 0.5,
        resourceId: "Alice",
        blockedBy: ["1"]
      },
      {
        id: "1",
        title: "Write tests for algorithms",
        start: todayForTest,
        end: new Date("2020-01-02"),
        duration: 2,
        position: 1,
        progress: 0.5,
        resourceId: "Bob"
      }
    ],
    todayForTest // optional
  )
);

// ->
[
  {
    id: "0",
    title: "Make scheduling algorithm",
    start: subDays(todayForTest, 2),
    end: new Date("2020-01-02"),
    duration: 4,
    position: 0,
    progress: 0.5,
    resourceId: "Alice",
    blockedBy: ["1"]
  },
  {
    id: "1",
    title: "Write tests for algorithms",
    start: addDays(todayForTest, 1),
    end: new Date("2020-01-03"),
    duration: 2,
    position: 1,
    progress: 0.5,
    resourceId: "Bob"
  }
];
```


CHANGELOG.md
# 1.1

- Add cyclic dependecies detection and fix
- Add toposort for output for better visualisation
- Numerous logic fixes and improvements

# 1.0

Initial release.


src/date.utils.test.ts
import { shiftToFirstNextBusinessDay } from "./date.utils";

describe("date.utils", () => {
  it("should shift +2 days if on saturday", () => {
    expect(shiftToFirstNextBusinessDay(new Date("2020-02-22"))).toEqual(
      new Date("2020-02-24")
    );
  });

  it("should shift +1 day if on sunday", () => {
    expect(shiftToFirstNextBusinessDay(new Date("2020-02-23"))).toEqual(
      new Date("2020-02-24")
    );
  });

  it("should not shift if on businessday", () => {
    expect(shiftToFirstNextBusinessDay(new Date("2020-02-24"))).toEqual(
      new Date("2020-02-24")
    );
    expect(shiftToFirstNextBusinessDay(new Date("2020-02-25"))).toEqual(
      new Date("2020-02-25")
    );
  });
});


src/graph.utils.test.ts
import { makeGraphFromTasks, Graph, makeReverseGraph } from "./graph.utils";
import { tasks } from "./mocks/tasks.mocks";

describe("graph.utils", () => {
  it("should make correct graph from tasks", () => {
    expect(makeGraphFromTasks(tasks)).toMatchSnapshot();
  });

  it("should make correct reverse graph", () => {
    const graph: Graph = new Map([
      ["0", new Set(["1", "3"])],
      ["1", new Set("2")],
      ["2", new Set("3")]
    ]);

    expect(makeReverseGraph(graph)).toMatchSnapshot();
  });
});


src/scheduleTasks.test.ts
import { scheduleTasks } from "./scheduleTasks";
import {
  tasks,
  correctlyScheduledTasks,
  todayForTest,
  tasksWithProgresses,
  correctlyScheduledTasksWithProgresses
} from "./mocks/tasks.mocks";

describe("scheduleTasks", () => {
  it("should re-schedule tasks with 0 progresses", () => {
    expect(scheduleTasks(tasks, todayForTest)).toEqual(correctlyScheduledTasks);
  });

  it("should correctly schedule tasks with 50 progress", () => {
    expect(scheduleTasks(tasksWithProgresses, todayForTest)).toEqual(
      correctlyScheduledTasksWithProgresses
    );
  });
});


src/date.utils.ts
import {
  isEqual,
  isSaturday as fnsIsSaturDay,
  isSunday as fnsIsSunday,
  max,
  subBusinessDays as fnsSubBusinessDays,
  addBusinessDays as fnsAddBusinessDays,
  subDays,
  addDays as fnsAddDays,
} from "date-fns";
import { Task } from "./scheduleTasks";

export const addBusinessDays = fnsAddBusinessDays;
export const addDays = fnsAddDays;
export const substructBusinessDays = fnsSubBusinessDays;
export const subtractDays = subDays;
export const maxDateTime = max;
export const getNowDate = () => new Date();

const isSaturday = fnsIsSaturDay;
const isSunday = fnsIsSunday;
const isDateEqual = (a: Date, b: Date) => isEqual(a, b);

/**
 * Update task dates, according to startDate change
 * forceUpdate param for intial set start-end dates according to duration, progress and business days
 *
 * @returns false if date didn't change, true if it changed
 */
export const updateTaskDatesByStart = (
  task: Task,
  newStartDate: Date,
  forceUpdate = false
): boolean => {
  const daysSpent = Math.floor(task.duration * task.progress);
  const newStartDateCorrectedByHolidays =
    shiftToFirstNextBusinessDay(newStartDate);
  const newStartDateCorrectedByProgress = substructBusinessDays(
    newStartDateCorrectedByHolidays,
    daysSpent
  );

  if (
    !forceUpdate &&
    isDateEqual(task.start, newStartDateCorrectedByProgress)
  ) {
    return false;
  }

  task.start = newStartDateCorrectedByProgress;

  // -1 because every task is minimum 1 day long,
  // say it starts and ends on same date, so it has 1 day duration
  task.end = addBusinessDays(
    newStartDateCorrectedByProgress,
    task.duration - 1
  );
  return true;
};

export const shiftToFirstNextBusinessDay = (date: Date) => {
  if (isSaturday(date)) {
    return addDays(date, 2);
  } else if (isSunday(date)) {
    return addDays(date, 1);
  } else {
    return date;
  }
};


src/graph.utils.ts
import { nullthrows } from "./misc";
import { TasksById, Graph, Task } from "./scheduleTasks";

type ID = string;

/**
 * Graph respects exlicit dependecies
 * and implicit by resources (via positions)
 */
export const makeGraphFromTasks = (tasks: Array<Task>): Graph => {
  // task and blockedBy
  const graph: Graph = new Map();
  const resourcesTasks = new Map<ID, Array<Task>>();

  // Create graphs
  for (const t of tasks) {
    // resource and its tasks
    const tasksOfResource = resourcesTasks.get(t.resourceId) ?? [];
    tasksOfResource.push(t);

    resourcesTasks.set(t.resourceId, tasksOfResource);

    graph.set(t.id, new Set(t.blockedBy ?? []));
  }

  // Now add deps
  for (const tasksOfResource of resourcesTasks.values()) {
    // first sort by position so links of tasks starts with higher position
    // then topological sort to reduce cyclic deps
    tasksOfResource.sort((a, b) => a.position - b.position);
    // is toposort needed?
    const sortedTasks = toposort(graph, tasksOfResource);

    // add to graph such edges so current node has prev as dependency (blocked by prev)
    let prevTask: Task | void;
    for (const task of sortedTasks) {
      if (
        prevTask &&
        prevTask.id !== task.id &&
        // has no incoming edge as well (otherwise it will be cyclic dep)
        !graph.get(prevTask.id)?.has(task.id)
      ) {
        graph.get(task.id)?.add(prevTask.id);
      }
      prevTask = task;
    }
  }

  return graph;
};

export const makeReverseGraph = (graph: Graph): Graph => {
  const reverseGraph: Graph = new Map();

  for (const [id, parentId] of dfs(graph, { withParentId: true })) {
    const prerequesitions = reverseGraph.get(id) ?? new Set();
    if (parentId !== undefined) {
      prerequesitions.add(parentId);
    }
    reverseGraph.set(id, prerequesitions);
  }

  return reverseGraph;
};

/**
 * Iterate over every node.
 * If withParentId = true, than it is possible to visit same not more then once
 * @yields {[string, string?]} [nodeId, parentNodeId?]
 */
export function* dfs(
  graph: Graph,
  options: { withParentId: boolean } = { withParentId: false }
): Generator<[string, string | void], void, void> {
  const visited = new Set<ID>();

  // DFS interative
  // iterate over all nodes in case of disconnected graph
  for (const node of graph.keys()) {
    if (visited.has(node)) {
      continue;
    }

    const stack: Array<ID> = [node];
    while (stack.length > 0) {
      const nodeId = stack.pop();
      nullthrows(nodeId);

      if (visited.has(nodeId)) {
        continue;
      }
      visited.add(nodeId);

      yield [nodeId, undefined];

      const blockedBy = graph.get(nodeId);
      if (!blockedBy) {
        continue;
      }
      for (const dependencyId of blockedBy) {
        if (options.withParentId) {
          // possible to yield same nodeId multiple times (needed for making reverse graph)
          yield [dependencyId, nodeId];
        }

        stack.push(dependencyId);
      }
    }
  }
}

/**
 * Main source of cyclic dependecies is previous step where graph is created
 * Often top-level task has same owner as children tasks
 * Since we create edge in graph also by same owner that's why there is cyclic deps
 *
 * IDEA: mitigate the issue by starting DFS walk from top-level (source) tasks!
 */
export const removeCyclicDependencies = (
  graph: Graph,
  tasks: Array<Task>
): void => {
  // Track visited to avoid computing path for already computed nodes
  const visited = new Set();
  let cyclicDepsRemovedCount = 0;

  const dfsAndRemove = (rootTaskId: ID) => {
    // [current task ID, set of previously visited tasks]
    const stack: Array<[ID, Set<ID>]> = [[rootTaskId, new Set()]];

    while (stack.length > 0) {
      const nextValue = stack.pop();
      nullthrows(nextValue);
      const [taskId, prevSet] = nextValue;

      const blockedBy = graph.get(taskId) ?? new Set();

      visited.add(taskId);

      for (const blockedById of blockedBy) {
        // cycle detected!
        if (prevSet.has(blockedById)) {
          // remove that edge
          blockedBy.delete(blockedById);
          cyclicDepsRemovedCount++;
          continue;
        }

        const newPrevSet = new Set(prevSet);
        newPrevSet.add(blockedById);
        stack.push([blockedById, newPrevSet]);
      }
    }
  };

  for (const task of tasks) {
    if (visited.has(task.id)) {
      continue;
    }

    dfsAndRemove(task.id);
  }

  console.debug("Cyclic deps removed:", cyclicDepsRemovedCount);
};

/**
 * Topological sort. DFS variant. TO-DO: try Kahn version
 * IMPORTANT! Graph shouldn't have cyclic deps.
 */
export const toposort = (graph: Graph, nodes: Array<Task>): Array<Task> => {
  const nodesById: TasksById = Object.fromEntries(nodes.map((t) => [t.id, t]));
  const visited = new Set<ID>();
  const order: Array<Task> = [];

  const dfs = (node: Task) => {
    const nodeId = node.id;
    const blockedBy = graph.get(nodeId) ?? new Set<ID>();

    if (visited.has(nodeId)) {
      return;
    }
    visited.add(nodeId);

    for (const idBlockedBy of blockedBy) {
      const nextNode = nodesById[idBlockedBy];
      if (nextNode) {
        dfs(nextNode);
      }
    }

    order.push(node);
  };

  for (const node of nodes) {
    if (visited.has(node.id)) {
      continue;
    }

    dfs(node);
  }

  return order;
};

/**
 * Iterate over every node setting nested levels
 * i.e. T1 (level 1) -> T2 (level 1.1) -> T3 (level 1.11)
 * @yields {[string, number, number]} [nodeId, nestingLevel, position]
 */
export function* dfsWithLevels(
  graph: Graph
): Generator<[string, number, number], void, void> {
  const visited = new Set<ID>();

  // DFS interative
  // iterate over all nodes in case of disconnected graph
  let topNodesCount = 1;
  for (const node of graph.keys()) {
    if (visited.has(node)) {
      continue;
    }

    const stack: Array<[ID, number, number]> = [[node, 1, topNodesCount++]];
    while (stack.length > 0) {
      const nextValue = stack.pop();
      nullthrows(nextValue);
      const [currentNodeId, currentNodeNesting, currentNodePosition] =
        nextValue;
      nullthrows(currentNodeId);

      yield [currentNodeId, currentNodeNesting, currentNodePosition];

      visited.add(currentNodeId);

      const blockedBy = graph.get(currentNodeId);
      if (!blockedBy) {
        continue;
      }

      let i = 1;
      for (const dependencyId of blockedBy) {
        if (visited.has(dependencyId)) {
          continue;
        }
        const nestingLevel = currentNodeNesting + 1;
        stack.push([
          dependencyId,
          nestingLevel,
          currentNodePosition + i * (1 / Math.pow(10, nestingLevel)),
        ]);

        i++;
      }
    }
  }
}


src/scheduleTasks.ts
import {
  makeGraphFromTasks,
  dfs,
  removeCyclicDependencies,
  toposort,
} from "./graph.utils";
import {
  addDays,
  getNowDate,
  maxDateTime,
  subtractDays,
  updateTaskDatesByStart,
} from "./date.utils";

export type ID = string;

export type Graph = Map<ID, Set<ID>>;
export type TasksById = { [id: string]: Task };
export type Task = {
  id: ID;
  title: string;
  start: Date;
  end: Date;
  duration: number;

  /**
   * Approximation of priority
   */
  position: number;
  progress: number;
  resourceId: ID;
  /**
   * Current task blocked by these tasks (depends on)
   */
  blockedBy?: Array<ID>;
};

export const scheduleTasks = (
  inputTasks: Array<Task>,
  today: Date = getNowDate()
): Array<Task> => {
  const dayBeforeToday = subtractDays(today, 1);
  const tasks: Array<Task> = inputTasks.map((t) => ({ ...t }));
  const tasksById: TasksById = Object.fromEntries(tasks.map((t) => [t.id, t]));
  const graph = makeGraphFromTasks(tasks);
  let cyclesToFullyUpdateDates = 1;

  // 1. Remove cyclic dependencies
  removeCyclicDependencies(graph, tasks);

  // 2. Initial update of all tasks start and ends days taking into account business days
  for (const task of tasks) {
    updateTaskDatesByStart(task, today, true);
  }

  // Repeat until dates remains unchanged, max graph.size times.
  // Similar to optimization in Bellman-Ford algorithm
  // @see https://en.wikipedia.org/wiki/Bellman–Ford_algorithm#Improvements
  for (let i = 0; i < tasks.length; i++) {
    let isAnyTaskTimelineChanged = false;

    for (const [taskId] of dfs(graph)) {
      const task = tasksById[taskId];
      // if blockedBy task not in initial data set
      if (task === undefined) {
        continue;
      }
      const blockedByTasks = Array.from(graph.get(task.id) ?? [])
        .map((blockedById) => tasksById[blockedById])
        // do not take into account tasks not in graph
        .filter(Boolean);
      const blockedByEndDates = blockedByTasks.map((t) => t.end);
      // add dayBeforeToday by default, so task without blockedBy starts on today
      blockedByEndDates.push(dayBeforeToday);

      // Start task on the next day after previous (blockedBy) tasks ends
      const maxBlockedByEndDate = addDays(maxDateTime(blockedByEndDates), 1);
      const isTaskTimelineUpdated = updateTaskDatesByStart(
        task,
        maxBlockedByEndDate
      );
      if (isTaskTimelineUpdated) {
        isAnyTaskTimelineChanged = true;
      }
    }

    if (isAnyTaskTimelineChanged === false) {
      break;
    }
    cyclesToFullyUpdateDates++;

    if (isAnyTaskTimelineChanged && i === tasks.length - 1) {
      console.error(
        'It\'s not enought "tasks.length" interations to fully schedule all tasks!'
      );
    }
  }

  console.debug(
    `Cycles to fully update dates ${cyclesToFullyUpdateDates}/${tasks.length}`
  );

  return toposort(graph, tasks);
};


src/mocks/tasks.mocks.ts
import { Task } from "../scheduleTasks";
import { subDays } from "date-fns";

export const todayForTest = new Date("2020-01-01");
export const tasks: Task[] = [
  {
    id: "0",
    title: "Make scheduling algorithm",
    start: todayForTest,
    end: new Date("2020-01-02"),
    duration: 2,
    position: 0,
    progress: 0,
    resourceId: "Alice",
    dependencies: ["1"]
  },
  {
    id: "1",
    title: "Write tests for algorithms",
    start: todayForTest,
    end: new Date("2020-01-02"),
    duration: 2,
    position: 1,
    progress: 0,
    resourceId: "Bob"
  },
  {
    id: "2",
    title: "Research a lot of gantt plotting libs",
    start: todayForTest,
    end: new Date("2020-01-02"),
    duration: 2,
    position: 2,
    progress: 0,
    resourceId: "Bob",
    dependencies: ["3"]
  },
  {
    id: "3",
    title: "Write your own",
    start: todayForTest,
    end: new Date("2020-01-02"),
    duration: 2,
    position: 3,
    progress: 0,
    resourceId: "Alice"
  }
];

export const correctlyScheduledTasks: Task[] = [
  {
    id: "0",
    title: "Make scheduling algorithm",
    start: todayForTest,
    end: new Date("2020-01-02"),
    duration: 2,
    position: 0,
    progress: 0,
    resourceId: "Alice",
    dependencies: ["1"]
  },
  {
    id: "1",
    title: "Write tests for algorithms",
    start: new Date("2020-01-03"),
    end: new Date("2020-01-06"),
    duration: 2,
    position: 1,
    progress: 0,
    resourceId: "Bob"
  },
  {
    id: "2",
    title: "Research a lot of gantt plotting libs",
    start: new Date("2020-01-07"),
    end: new Date("2020-01-08"),
    duration: 2,
    position: 2,
    progress: 0,
    resourceId: "Bob",
    dependencies: ["3"]
  },
  {
    id: "3",
    title: "Write your own",
    start: new Date("2020-01-09"),
    end: new Date("2020-01-10"),
    duration: 2,
    position: 3,
    progress: 0,
    resourceId: "Alice"
  }
];

export const tasksWithProgresses: Task[] = [
  {
    id: "0",
    title: "Make scheduling algorithm",
    start: todayForTest,
    end: new Date("2020-01-04"),
    duration: 4,
    position: 0,
    progress: 0.5,
    resourceId: "Alice",
    dependencies: ["1"]
  },
  {
    id: "1",
    title: "Write tests for algorithms",
    start: todayForTest,
    end: new Date("2020-01-02"),
    duration: 2,
    position: 1,
    progress: 0.5,
    resourceId: "Bob"
  }
];

export const correctlyScheduledTasksWithProgresses: Task[] = [
  {
    id: "0",
    title: "Make scheduling algorithm",
    start: subDays(todayForTest, 2),
    end: new Date("2020-01-02"),
    duration: 4,
    position: 0,
    progress: 0.5,
    resourceId: "Alice",
    dependencies: ["1"]
  },
  {
    id: "1",
    title: "Write tests for algorithms",
    start: new Date("2020-01-02"),
    end: new Date("2020-01-03"),
    duration: 2,
    position: 1,
    progress: 0.5,
    resourceId: "Bob"
  }
];
